common/scala/src/main/scala/whisk/core/entity/InvokerInstanceId.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.entity
+
+import spray.json.DefaultJsonProtocol
+
+case class InvokerInstanceId(val instance: Int, name: Option[String] = None) {
+  def toInt: Int = instance
+}
+
+case class ControllerInstanceId(private val instance: String) {
+  //keep instance private, since we will replace illegal chars for kafka usage
+  //for kafka topic legal chars see https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/internals/Topic.java#L29
+  val ILLEGAL_CHARS = "[^a-zA-Z0-9._-]".r
+  //validate once on construction - reserve 20 (arbitrary) chars for prefix to be prepended to create topic names
+  private val MAX_NAME_LENGTH = 249 - 20
+  val replaced = ILLEGAL_CHARS.replaceAllIn(instance, "")
+  require(replaced.length <= MAX_NAME_LENGTH, s"topic name can be at most ${MAX_NAME_LENGTH} ${replaced}")
+
+  override def toString() = replaced
```

Comment:
Please don't use `toString` for places where you programmatically need a value to be a `String`. It should be used for logging only.

We have a convention of using `asString` for places where you really need a string in code.

Reply:
ok, changing the refs to use `asString`. Note that the log markers are also changing to use this, not sure if that is a problem, but I think it is appropriate.